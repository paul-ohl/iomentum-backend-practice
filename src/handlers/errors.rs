use serde::Serialize;
use serde_json::json;
use warp::{
    http::StatusCode,
    reject::Rejection,
    reply::{self, Reply},
};

use crate::domain::errors::{Error, Result};

pub fn result_to_warp_reply<T>(result: Result<T>) -> Result<impl Reply, Rejection>
where
    T: Serialize,
{
    match result {
        Ok(data) => Ok(reply::with_status(reply::json(&data), StatusCode::ACCEPTED)),

        Err(e) => {
            let message = e.to_string();
            Ok(reply::with_status(
                reply::json(&json!({
                    "status": "fail",
                    "message": message,
                })),
                to_http_status_code(e),
            ))
        }
    }
}

fn to_http_status_code(e: Error) -> StatusCode {
    match e {
        Error::InternalError(_) => StatusCode::INTERNAL_SERVER_ERROR,
        Error::UsernameAlreadyExists => StatusCode::CONFLICT,
        Error::UserNotFound => StatusCode::NO_CONTENT,
        Error::UserCreationFailed(_) => StatusCode::INTERNAL_SERVER_ERROR,
        Error::UserFetchFailed(_) => StatusCode::INTERNAL_SERVER_ERROR,
        Error::UserUpdateFailed(_) => StatusCode::INTERNAL_SERVER_ERROR,
        Error::UserDeletionFailed(_) => StatusCode::INTERNAL_SERVER_ERROR,
        Error::TicketFetchFailed(_) => StatusCode::INTERNAL_SERVER_ERROR,
        Error::TicketNotFound => StatusCode::NO_CONTENT,
        Error::TicketCreationFailed(_) => StatusCode::INTERNAL_SERVER_ERROR,
        Error::TicketUpdateFailed(_) => StatusCode::INTERNAL_SERVER_ERROR,
        Error::TicketDeletionFailed(_) => StatusCode::INTERNAL_SERVER_ERROR,
        Error::InvalidPassword(_) => StatusCode::BAD_REQUEST,
        Error::InvalidUsername(_) => StatusCode::BAD_REQUEST,
        Error::InvalidRole(_) => StatusCode::BAD_REQUEST,
        Error::LoginFailed(_) => StatusCode::UNAUTHORIZED,
    }
}
