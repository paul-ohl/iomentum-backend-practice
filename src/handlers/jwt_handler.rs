use std::collections::BTreeMap;

use hmac::{Hmac, Mac};
use jwt::{SignWithKey, VerifyWithKey};
use sha2::Sha256;

use crate::domain::{
    errors::{Error, Result},
    types::JwtClaims,
};

pub struct JwtHandler {
    key: Hmac<Sha256>,
}

impl JwtHandler {
    pub fn new(secret: String) -> Result<Self> {
        Ok(Self {
            key: Hmac::new_from_slice(secret.as_bytes())
                .expect("Failed to create HMAC from secret:"),
        })
    }

    pub fn generate_token(&self, claims: JwtClaims) -> Result<String> {
        claims
            .to_tree_map()
            .sign_with_key(&self.key)
            .map_err(|e| Error::InternalError(format!("Failed to sign token: {}", e)))
    }

    pub fn decode_token(&self, token: &str) -> Result<JwtClaims> {
        let tree: BTreeMap<String, String> = token
            .verify_with_key(&self.key)
            .map_err(|e| Error::LoginFailed(format!("Couldn't decode JWT: {e}")))?;
        tree.try_into()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;
    use uuid::Uuid;

    #[test]
    fn test_generate_token() {
        let jwt_handler = JwtHandler::new("secret".to_string()).unwrap();
        let claims = JwtClaims::new("username".to_string(), Uuid::new_v4(), "role".to_string());
        let token = jwt_handler.generate_token(claims).unwrap();
        assert!(!token.is_empty());
    }

    #[test]
    fn test_decode_token() {
        let jwt_handler = JwtHandler::new("secret".to_string()).unwrap();
        let claims = JwtClaims::new("username".to_string(), Uuid::new_v4(), "role".to_string());
        let token = jwt_handler.generate_token(claims).unwrap();
        let decoded_claims = jwt_handler.decode_token(&token).unwrap();
        assert_eq!(decoded_claims.sub, "username");
        assert_eq!(decoded_claims.role, "role");
        assert!(decoded_claims.exp > Utc::now().timestamp());
    }
}
