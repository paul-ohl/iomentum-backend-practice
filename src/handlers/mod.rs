pub mod errors;
pub mod password_hasher;
pub mod tickets;
pub mod users;
