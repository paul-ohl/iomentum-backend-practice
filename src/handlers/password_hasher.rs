use argon2::{
    password_hash::SaltString, Argon2, PasswordHash, PasswordHasher as ArgonPasswordHasher,
    PasswordVerifier,
};
use rand_core::OsRng;

pub fn hash_password(password: &str) -> Result<String, String> {
    let salt = SaltString::generate(&mut OsRng);

    let password_hash = Argon2::default()
        .hash_password(password.as_bytes(), &salt)
        .map_err(|_| {
            "Failed to hash password, please try again or contact support if the problem persists"
                .to_string()
        })?
        .to_string();
    Ok(password_hash)
}

pub fn verify(password_sent: &str, db_hash: &str) -> bool {
    PasswordHash::new(db_hash).map_or(false, |parsed_hash| {
        Argon2::default()
            .verify_password(password_sent.as_bytes(), &parsed_hash)
            .is_ok()
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hash_password() {
        let password = "password";
        let hashed_password = hash_password(password).unwrap();
        assert_ne!(password, hashed_password);
    }

    #[test]
    fn test_verify() {
        let password = "password";
        let hashed_password = hash_password(password).unwrap();
        assert!(verify(password, &hashed_password));
    }
}
