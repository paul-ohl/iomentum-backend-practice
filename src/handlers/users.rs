use std::sync::Arc;

use sqlx::PgPool;
use warp::{reject::Rejection, reply::Reply};

use crate::{
    domain::dtos::user_dtos::UserInputDto, handlers::errors::result_to_warp_reply, models::users,
};

type ReplyRes<T> = Result<T, Rejection>;

pub async fn get_all_users(db_pool: Arc<PgPool>) -> ReplyRes<impl Reply> {
    let users = users::get_all(db_pool).await;
    result_to_warp_reply(users)
}

pub async fn get_user_by_id(id: uuid::Uuid, db_pool: Arc<PgPool>) -> ReplyRes<impl Reply> {
    let user = users::get_by_id(db_pool, id).await;
    result_to_warp_reply(user)
}

pub async fn get_user_by_username(username: String, db_pool: Arc<PgPool>) -> ReplyRes<impl Reply> {
    let users = users::get_by_username(db_pool, username).await;
    result_to_warp_reply(users)
}

pub async fn update_user(
    id: uuid::Uuid,
    user_input: UserInputDto,
    db_pool: Arc<PgPool>,
) -> ReplyRes<impl Reply> {
    match user_input.try_into() {
        Ok(user_modifications) => {
            result_to_warp_reply(users::update(db_pool, id, user_modifications).await)
        }
        Err(e) => result_to_warp_reply(Err(e)),
    }
}

pub async fn delete_user(id: uuid::Uuid, db_pool: Arc<PgPool>) -> ReplyRes<impl Reply> {
    let user_id = users::delete(db_pool, id).await;
    result_to_warp_reply(user_id)
}

pub async fn register_user(user_input: UserInputDto, db_pool: Arc<PgPool>) -> ReplyRes<impl Reply> {
    match user_input.try_into() {
        Ok(new_user) => result_to_warp_reply(users::create(db_pool, new_user).await),
        Err(e) => result_to_warp_reply(Err(e)),
    }
}
