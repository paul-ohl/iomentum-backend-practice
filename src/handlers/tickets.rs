use std::sync::Arc;

use sqlx::PgPool;
use uuid::Uuid;
use warp::{reject::Rejection, reply::Reply};

use crate::models::tickets::{self, TicketInput};

use super::errors::result_to_warp_reply;

type ReplyRes<T> = Result<T, Rejection>;

pub async fn get_all_tickets(db: Arc<PgPool>) -> ReplyRes<impl Reply> {
    let tickets = tickets::get_all(db).await;
    result_to_warp_reply(tickets)
}

pub async fn get_ticket_by_id(id: Uuid, db: Arc<PgPool>) -> ReplyRes<impl Reply> {
    let ticket = tickets::get_by_id(db, id).await;
    result_to_warp_reply(ticket)
}

pub async fn get_ticket_by_user_id(user_id: Uuid, db: Arc<PgPool>) -> ReplyRes<impl Reply> {
    let tickets = tickets::get_by_user(db, user_id).await;
    result_to_warp_reply(tickets)
}

pub async fn create_ticket(ticket: TicketInput, db: Arc<PgPool>) -> ReplyRes<impl Reply> {
    let ticket_id = tickets::create(db, ticket).await;
    result_to_warp_reply(ticket_id)
}

pub async fn update_ticket(
    id: Uuid,
    ticket_input: TicketInput,
    db: Arc<PgPool>,
) -> ReplyRes<impl Reply> {
    let ticket_id = tickets::update(db, id, ticket_input).await;
    result_to_warp_reply(ticket_id)
}

pub async fn delete_ticket(id: Uuid, db: Arc<PgPool>) -> ReplyRes<impl Reply> {
    let ticket_id = tickets::delete_one(db, id).await;
    result_to_warp_reply(ticket_id)
}
