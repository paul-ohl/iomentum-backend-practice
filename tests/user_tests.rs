mod helper;
use helper::spawn_app;
use iomentum_backend_practice::handlers::password_hasher;
use serde_json::json;

#[tokio::test]
async fn crud_users_works() {
    let test_app = spawn_app().await;
    let client = reqwest::Client::new();

    let response = client
        .post(format!("{}/register", test_app.address))
        .body(
            json!({
                "username": "test1",
                "password": "test1234",
                "role": "user",
            })
            .to_string(),
        )
        .send()
        .await
        .expect("Failed to execute request.");

    assert!(response.status().is_success());
    let saved = sqlx::query!("SELECT id, username, password_hash, role FROM Users")
        .fetch_one(&test_app.db_pool)
        .await
        .expect("Failed to fetch from db.");
    assert_eq!("test1", saved.username);
    assert_eq!("user", saved.role);
    assert!(password_hasher::verify("test1234", &saved.password_hash));
    let id = saved.id;

    // get user by id
    let response = client
        .get(format!("{}/users/{}", test_app.address, id))
        .send()
        .await
        .expect("Failed to execute request.");
    assert!(response.status().is_success());
    assert_eq!(
        "test1",
        response.json::<serde_json::Value>().await.unwrap()["username"]
    );

    let response = client
        .post(format!("{}/register", test_app.address))
        .body(
            json!({
                "username": "test2",
                "password": "2short",
                "role": "user",
            })
            .to_string(),
        )
        .send()
        .await
        .expect("Failed to execute request.");
    assert!(!response.status().is_success());
    insta::assert_json_snapshot!(response.json::<serde_json::Value>().await.unwrap());

    // get all users
    let response = client
        .get(format!("{}/users", test_app.address))
        .send()
        .await
        .expect("Failed to execute request.");
    assert!(response.status().is_success());
    assert_eq!(
        id.to_string(),
        response.json::<serde_json::Value>().await.unwrap()[0]["id"]
    );

    // get user by username
    let response = client
        .get(format!("{}/users/by-username/test1", test_app.address))
        .send()
        .await
        .expect("Failed to execute request.");
    assert!(response.status().is_success());
    let data = response.json::<serde_json::Value>().await.unwrap();
    assert_eq!(id.to_string(), data["id"]);
    assert_eq!("test1", data["username"]);

    // update user
    let response = client
        .patch(format!("{}/users/{}", test_app.address, id))
        .body(
            json!({
                "username": "test3",
                "password": "newpassword",
                "role": "admin",
            })
            .to_string(),
        )
        .send()
        .await
        .expect("Failed to execute request.");
    assert!(response.status().is_success());
    let saved = sqlx::query!("SELECT id, username, password_hash, role FROM Users")
        .fetch_one(&test_app.db_pool)
        .await
        .expect("Failed to fetch from db.");
    assert_eq!("test3", saved.username);
    assert_eq!("admin", saved.role);
    assert!(password_hasher::verify("newpassword", &saved.password_hash));
}
