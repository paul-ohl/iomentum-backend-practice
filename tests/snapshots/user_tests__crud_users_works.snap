---
source: tests/user_tests.rs
expression: "response.json::<serde_json::Value>().await.unwrap()"
---
{
  "message": "invalid password: Password must be at least 8 characters long",
  "status": "fail"
}
